nirum_wsgi.py

```diff
@@ -137,106 +168,128 @@ def __call__(self, environ, start_response):
         """
         return self.route(environ, start_response)
 
-    def route(self, environ, start_response):
-        """Route
-
-        :param environ:
-        :param start_response:
-
-        """
+    def dispatch_method(self, environ):
+        payload = None
         request = Request(environ)
         service_methods = self.service.__nirum_service_methods__
-        error_raised = None
-        for _, pattern, verb, method_name in self.rules:
-            path_info = environ['PATH_INFO']
-            if isinstance(path_info, bytes):
-                # FIXME Decode properly; URI is not unicode
-                path_info = path_info.decode()
-            match = pattern.match(path_info)
-            if match and environ['REQUEST_METHOD'] == verb.upper():
-                routed = True
-                service_method = method_name
-                if verb in ('GET', 'DELETE'):
-                    method_parameters = {
-                        k: v
-                        for k, v in service_methods[method_name].items()
-                        if not k.startswith('_')
-                    }
-                    # TODO Parsing query string
-                    payload = {p: match.group(p) for p in method_parameters}
-                else:
-                    try:
-                        payload = parse_json_payload(request)
-                    except InvalidJsonError as e:
-                        error_raised = self.error(
+        # CORS
+        cors_headers = [('Vary', 'Origin')]
+        request_match, matched_verb = match_request(
+            self.rules, request.path, request.method
+        )
+        if request_match:
+            service_method = request_match.method_name
+            cors_headers.append(
+                (
+                    'Access-Control-Allow-Methods',
+                    ', '.join(matched_verb + ['OPTIONS'])
+                )
+            )
+            method_parameters = {
+                k: v
+                for k, v in service_methods[request_match.method_name].items()
+                if not k.startswith('_')
+            }
+            payload = {
+                p: request_match.match_group.group(p)
+                for p in method_parameters
+            }
+            # TODO Parsing query string
+            if request_match.verb not in ('GET', 'DELETE'):
+                try:
+                    json_payload = parse_json_payload(request)
+                except InvalidJsonError as e:
+                    raise MethodNoMatchError(
```

Comment:
It would better if the exception class takes only a `request` and a `message` string rather than a whole response object.  We could make an error response from an exception value when it's caught.